packages/amplify-codegen-e2e-tests/src/codegen-tests-base/add-codegen.ts

```diff
@@ -11,12 +11,122 @@ import {
+  statementNamePattern?: string;
+  maxDepth?: number;
+  isTypeGenerated?: boolean;
+  typeFileName?: string;
+}
+
+const defaultConfig: CodegenMatrixTestProps = {
+  ...DEFAULT_JS_CONFIG,
+  params: [],
+  isAPINotAdded: false,
+  isCodegenAdded: false,
+  isStatementGenerated: true,
+  statementNamePattern: '\r', // default value
+  maxDepth: 2, // default value
+  isTypeGenerated: true,
+  typeFileName: '\r', // default value
+}
+
+/**
+ * Extract type file path given the test props. When the carriage return is detected, the default path will be returned
+ * @param props codegen matrix test props
+ * @param projectRoot project root
+ * @returns type file path
+ */
+const getTypeFilePath = (props: CodegenMatrixTestProps, projectRoot: string): string | undefined => {
+  if (props.typeFileName) {
+    if (props.typeFileName === '\r') {
+      if (props.frontendType === AmplifyFrontend.ios) {
+        return path.join(projectRoot, 'API.swift')
+      } else if (props.frontendType === AmplifyFrontend.javascript) {
+        switch (props.codegenTarget) {
+          case 'flow':
+            return path.join(projectRoot, 'src', 'API.js');
+          case 'angular':
+            return path.join(projectRoot, 'src', 'app', 'API.service.ts');
+          case 'typescript':
+            return path.join(projectRoot, 'src', 'API.ts');
+          default:
+            return undefined;
+        }
+      }
+    }
+    return path.join(projectRoot, props.typeFileName);
+  }
+  return undefined;
+}
+
+/**
+ * Check if the type file(s) is generated at the given file path(single file or directory)
+ * @param filePath type file path
+ * @returns is type file generated or not
+ */
+const isTypeFileGeneratedAtPath = (filePath: string | undefined): boolean => {
+  if (filePath) {
+    if (existsSync(filePath)) {
+      // if path is directory, check if files are generated within
+      if (lstatSync(filePath).isDirectory()) {
+        return readdirSync(filePath).length > 0;
+      }
+      // path is a single file
+      return true;
+    }
+    return false;
+  }
+  return false;
+}
+
+/**
+ * Util function for testing codegen matrix
+ * @param props codegen matrix test properties
+ * @param projectRoot project root
+ */
+export async function testAddCodegenMatrix(props: CodegenMatrixTestProps, projectRoot: string): Promise<void> {
+  const config: CodegenMatrixTestProps = { ...defaultConfig, ...props }
+  if (config.graphqlCodegenDir) {
+    deleteProjectDir(path.join(projectRoot, config.graphqlCodegenDir));
+  }
+  if (config.srcDir !== '.') {
+    deleteProjectDir(path.join(projectRoot, config.srcDir));
+  }
+  const typeFilePath = getTypeFilePath(config, projectRoot);
+  if (typeFilePath && existsSync(typeFilePath)) {
+    rmSync(typeFilePath);
+  }
+  if (config.framework === 'angular') {
+    // Mock angular json file otherwise the project configure will fail
+    writeFileSync(path.join(projectRoot, 'angular.json'), '{}');
+  }
```

Comment:
This is to mock a `angular.json` file at project root. Otherwise the frontend type change will fail.